src/app/InteractionModelEngine.cpp

```diff
@@ -764,34 +860,131 @@ CHIP_ERROR InteractionModelEngine::CanEstablishReadTransaction(const ReadHandler
+
+    if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+        usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
     {
-        // Doesn't matter how many other reads are in progress; we are not
-        // going to handle this request.
-        return (allowUnlimited ? CHIP_NO_ERROR : CHIP_IM_GLOBAL_STATUS(PathsExhausted));
+        // We have enough resources, then we serve the requests in a best-effort manner.
+        return Protocols::InteractionModel::Status::Success;
     }
 
-    mReadHandlers.ForEachActiveObject([&](ReadHandler * handler) {
-        if (handler->GetAccessingFabricIndex() == currentFabricIndex && handler->IsType(ReadHandler::InteractionType::Read))
+    if ((aRequestedAttributePathCount > kMinSupportedPathsPerReadRequest &&
+         usedAttributePaths + aRequestedAttributePathCount > attributePathCap) ||
+        (aRequestedEventPathCount > kMinSupportedPathsPerReadRequest && usedEventPaths + aRequestedEventPathCount > eventPathCap))
+    {
+        // We cannot offer enough resources, and the read transaction is requesting more than the spec limit.
+        return Protocols::InteractionModel::Status::PathsExhausted;
+    }
+
+    // If we have commissioned CHIP_CONFIG_MAX_FABRICS already, and this transaction don't have an associated fabric index, reject
+    // the request if we don't have sufficient resources for this request.
+    if (mpFabricTable->FabricCount() == GetConfigMaxFabrics() && aFabricIndex == kUndefinedFabricIndex)
+    {
+        return Protocols::InteractionModel::Status::ResourceExhausted;
+    }
+
+    size_t usedAttributePathsInFabric = 0;
+    size_t usedEventPathsInFabric     = 0;
+    size_t usedReadHandlersInFabric   = 0;
+    mReadHandlers.ForEachActiveObject([&](auto * handler) {
+        if (!handler->IsType(ReadHandler::InteractionType::Read) || handler->GetAccessingFabricIndex() != aFabricIndex)
         {
-            activeReadHandlersOnCurrentFabric++;
+            return Loop::Continue;
         }
+        usedAttributePathsInFabric += handler->GetAttributePathCount();
+        usedEventPathsInFabric += handler->GetEventPathCount();
+        usedReadHandlersInFabric++;
         return Loop::Continue;
     });
 
-    // The incoming read handler here is also counted above.
-    if (activeReadHandlersOnCurrentFabric > kReservedReadHandlersPerFabricForReadRequests)
+    // Resources exhausted, since there is already some read requests ongoing on this fabric, please retry later.
+    if (usedAttributePathsInFabric + aRequestedAttributePathCount >
+            kMinSupportedPathsPerReadRequest * kMinSupportedReadRequestsPerFabric ||
+        usedEventPathsInFabric + aRequestedEventPathCount > kMinSupportedPathsPerReadRequest * kMinSupportedReadRequestsPerFabric)
     {
-        return (allowUnlimited ? CHIP_NO_ERROR : CHIP_IM_GLOBAL_STATUS(Busy));
+        return Protocols::InteractionModel::Status::PathsExhausted;
```

Comment:
That says "retry with fewer paths", but that should only happen if this read was over the per-read path minimum.  The normal "valid number of paths, but we can't handler it right now" code is BUSY.